### DISCUSSION TITLE
The result of getServerSnapshot should be cached to avoid an infinite loop

### DISCUSSION THREAD

**Discussion by justrealmilk** [pmndrs/zustand#2855] [Bug report]
### Bug Description

Can someone explain to me why after updating to Next 15 and Zustand 5, I am required to wrap every single one of my Zustand hooks with useShallow? Is this intended? Is this experience by design? I want to scream

### Reproduction Link

-

**Comment by dbritto-dev** [comment]
Well, actually do you have many options: `createWithEqualityFn`, `useStoreWithEqualityFn`, and `useShallow` you can use whatever you want.

**Reply by justrealmilk** [reply]
Yeah but why do I need to update every single hook? I rolled back to 14 and it works fine with Next 15 (or so it seems)

**Reply by dbritto-dev** [reply]
@justrealmilk seems you didn't read the docs but you don't need to update every single hook, it depends on what you want to use

**Reply by EricWVGG** [reply]
Yeah, this doesn't really make much sense… it’s fine and all to say “read the docs” but this is a thing that worked yesterday and doesn’t work today, I’d like to know what exactly happened.

**Reply by dai-shi** [reply]
Even with Zustand v4, while it doesn't error, it doesn't optimize rerenders. So, v4 hides the potential issue, while v5 reveals it. `useShallow` fixes both cases.

See: https://stackblitz.com/edit/vitejs-vite-neuatc?file=src%2FApp.tsx&terminal=dev

**Reply by EricWVGG** [reply]
I’m not using computed state, though.

```
import { create } from 'zustand'

const useLayout = create<{
  navActive: boolean
  setNavActive: (b?: boolean) => void
}>((set) => ({
  navActive: false,
  setNavActive: (navActive) => set({ navActive }),
}))

export const NavButton = () => {
  const setNavActive = useLayout((state) => state.setNavActive)
  return <div onClick={() => setNavActive(true)}>🍔</div>
}
```

**Reply by dai-shi** [reply]
That sounds unexpected. The code looks okay.
Can you create a small reproduction with https://stackblitz.com ?

**Reply by justrealmilk** [reply]
For anyone interested in the aforementioned docs which I didn't read but could have been simply linked to https://zustand.docs.pmnd.rs/migrations/migrating-to-v5#requiring-stable-selector-outputs

It's way too much work to adjust every single hook. I'd rather look at alternative libraries than go to all that effort so I'll just stay at version 4 

Thanks!

**Reply by thomasters-nbl** [reply]
Such a frustrating change. I am also sticking at V4. What a nightmare. And what was with the unnecessarily unhelpful replies?

**Reply by GusakovskayaDev** [reply]
If you pull each value individually from the store, you can do without useShallow.:
```
const addItem = useStore(state => state.addItem);
const loading = useStore(state => state.loading);
```

But if you need multiple values, then you have to use useShallow for surface memoization.:
```
const [addItem, loading] = useStore(useShallow(state => [
    state.addItem, state.loading
  ]));
```
